src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -0,0 +1,443 @@
+                    (None, None) => {
+                        // Both cursors yield `None`: we know that there is nothing remaining.
+                        return Ok(None);
+                    }
+                    (None, Some(k)) => {
+                        // Write cursor yields `None` but lock cursor yields something:
+                        // In RC, it means we got nothing.
+                        // In SI, we need to check if the lock will cause conflict.
+                        (k, false, true)
+                    }
+                    (Some(k), None) => {
+                        // Write cursor yields something but lock cursor yields `None`:
+                        // We need to further step write cursor to our desired version
+                        (Key::truncate_ts_for(k)?, true, false)
+                    }
+                    (Some(wk), Some(lk)) => {
+                        let write_user_key = Key::truncate_ts_for(wk)?;
+                        match write_user_key.cmp(lk) {
+                            Ordering::Less => {
+                                // Write cursor user key < lock cursor, it means the lock of the
+                                // current key that write cursor is pointing to does not exist.
+                                (write_user_key, true, false)
+                            }
+                            Ordering::Greater => {
+                                // Write cursor user key > lock cursor, it means we got a lock of a
+                                // key that does not have a write. In SI, we need to check if the
+                                // lock will cause conflict.
+                                (lk, false, true)
+                            }
+                            Ordering::Equal => {
+                                // Write cursor user key == lock cursor, it means the lock of the
+                                // current key that write cursor is pointing to *exists*.
+                                (lk, true, true)
+                            }
+                        }
+                    }
+                };
+
+                // Use `from_encoded_slice` to reserve space for ts, so later we can append ts to
+                // the key or its clones without reallocation.
+                (Key::from_encoded_slice(res.0), res.1, res.2)
+            };
+
+            // Attempt to read specified version of the key. Note that we may get `None`
+            // indicating that no desired version is found, or a DELETE version is found, or
+            // when `has_write == false`. We may also get `Err` due to key lock. We need to
+            // ensure that the scanner keeps working for future calls even when meeting key
+            // lock errors (why??), so we don't apply `?` operator to the result here thus
+            // iterators can be moved forward.
+            let result = self.get(&current_user_key, has_write, has_lock);
```

Comment:
I think you don't need to push down `has_write` and `has_lock` into `get`, the bellow code is more clear:
```
let val = None;
if has_lock {
   check_lock ...
   move cursor to next key
}
if has_write {
   val = self.get(&cur_user_key);
   move cursor to next key
}
if let Some(v) = val {
   ...
}
```